src/client/Microsoft.Identity.Client/ManagedIdentity/ImdsManagedIdentitySource.cs

```diff
@@ -0,0 +1,167 @@
+        private readonly string _resourceId;
+        private readonly Uri _imdsEndpoint;
+
+        
+
+        internal ImdsManagedIdentitySource(RequestContext requestContext) : base(requestContext)
+        {
+            if (!string.IsNullOrEmpty(EnvironmentVariables.PodIdentityEndpoint))
+			{
+                requestContext.Logger.Verbose("[Managed Identity] Environment variable for IMDS returned endpoint: " + EnvironmentVariables.PodIdentityEndpoint);
+                var builder = new UriBuilder(EnvironmentVariables.PodIdentityEndpoint)
+                {
+                    Path = ImdsTokenPath
+                };
+                _imdsEndpoint = builder.Uri;
+			}
+			else
+			{
+                requestContext.Logger.Verbose("[Managed Identity] Unable to find an endpoint in environment variable for IMDS, using the default endpoint.");
+            	_imdsEndpoint = s_imdsEndpoint;
+			}
+
+            _clientId = requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedClientId;
+            _resourceId = requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedResourceId;
+        }
+
+        protected override ManagedIdentityRequest CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeHelper.ScopesToResource(scopes);
+
+            ManagedIdentityRequest request = new(HttpMethod.Get, _imdsEndpoint);
+
+            request.Headers.Add("Metadata", "true");
+            request.QueryParameters["api-version"] = ImdsApiVersion;
+            request.QueryParameters["resource"] = resource;
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                _requestContext.Logger.Verbose("[Managed Identity] Adding user assigned client id to the request.");
+                request.QueryParameters[Constants.ManagedIdentityClientId] = _clientId;
+            }
+            if (!string.IsNullOrEmpty(_resourceId))
+            {
+                _requestContext.Logger.Verbose("[Managed Identity] Adding user assigned resource id to the request.");
+                request.QueryParameters[Constants.ManagedIdentityResourceId] = _resourceId;
+            }
+
+            request.ComputeUri();
+            return request;
+        }
+
+        public override async Task<ManagedIdentityResponse> AuthenticateAsync(AppTokenProviderParameters parameters, CancellationToken cancellationToken)
+        {
+            try
+            {
+                return await base.AuthenticateAsync(parameters, cancellationToken).ConfigureAwait(false);
+            }
+            catch (MsalServiceException e) when (e.ErrorCode == MsalError.ManagedIdentityRequestFailed)
+            {
+                _requestContext.Logger.Error(NoResponseError + e.Message);
+                throw;
+            }
+            catch (TaskCanceledException)
+            {
+                _requestContext.Logger.Error(TimeoutError);
+                throw;
+            }
```

Comment:
Can this be moved to wrap the HTTP calls in `ManagedIdentitySource.AuthenticateAsync` instead?